proposals/context.md

```diff
@@ -0,0 +1,265 @@
+
+If we have to actually resolve a conflict in an application then a 'MappingContextProvider' is theorized. This provider would receive context events and remap them to namespaced context events then emit them up the tree. In this manner we can map or reroute context events easily.
+
+A possible implementation of this will be added to the `lit-labs` project soon.
+
+## Should requesters get to 'accept' providers?
+
+The current API as proposed does not allow a requestor to 'approve' that a provider is going to give it the right object. We have some capability to enforce this in Typescript, but we could provide a slightly different API that would allow the requesting component to check the value it will receive:
+
+```js
+this.dispatchEvent(
+  new RequestContextEvent('logger', (candidate) => {
+    if (typeof candidate.log === 'function' && typeof candidate.info === 'function) {
+      // we can accept this candidate
+      return (logger, dispose) => {
+        this.logger = logger;
+      };
+    }
+  });
+)
+```
+
+In this alternative, we expect to always synchronously receive a 'candidate' data value that a provider may give us. We can then inspect this candidate value in our component to determine if it has the correct shape, and then return our callback function to accept updates to it.
+
+In this proposal we would likely enforce that the callback always be invoked synchronously.
+
+Alternative APIs could also be explored in this approach, we could for instance have providers append themselves to a list of potential providers along with candidate value objects, and then allow our components to pick which provider they wish to use:
+
+```js
+const contextRequest = new RequestContextEvent("logger");
+this.dispatchEvent(context);
+if (!contextRequest.providers) {
+  // no providers for logger
+  return;
+}
+const provider = contextRequest.providers.find((loggerProvider) => {
+  // test if the provider is the type we want or the value it provides is right
+});
+const dispose = provider.provide(this, (logger) => {
+  this.logger = logger;
+});
+// later...
+dispose(); // don't need updates anymore
+```
+
+These alternatives do provide more capability, but its an open question as to whether or not this complexity is warranted or desired. It also opens up a larger question about what would the candidate value be? Would it have to be an object of the requested type, could it be some other protocol to determine uniformity between the requested data and the actual data? This begins to seem more complex than we really need here for unnecessary type safety overhead. It is suggested if consumers want type safety then they should use Typescript to achieve this.
```

Comment:
Have we seen a real use-case here yet or an analogy in a framework? I agree that implementing a runtime-typecheck callback seems overly complex. If really needed, this would be an argument for more guaranteed unique context keys.

Reply:
I agree there isn't a strong use case for these yet, and with the tighter binding to use context value objects rather than strings I think this helps avoid the need for validation?